### ISSUE TITLE
Getting  "undefined method user" 500 error when trying to do a create 

### ISSUE THREAD

**Issue by aeabbott** [ga-wdi-boston/full-stack-project#729]
Problem: I am trying to create a birthday that is linked to the current user when they are logged in.
I have ran the curl requests successfully to sign up a user and sign in that same user. When I use the token returned from the sign in user in my create birthday curl request, I am getting an undefined method user error

Here are my curl requests
```
$ curl "${API}${URL_PATH}" \
  --include \
   --request POST \
   --header "Content-Type: application/json" \
   --data '{
     "credentials": {
       "email": "amandabanana@yahoo.com",
       "password": "pass",
       "password_confirmation": "pass"
     }
   }'
HTTP/1.1 200 OK
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Type: application/json; charset=utf-8
ETag: W/"9c5abe5a414ff9c6c611404f8df5a59b"
Cache-Control: max-age=0, private, must-revalidate
X-Request-Id: 6fe07e32-4716-4440-9207-32bf36cffdbd
X-Runtime: 0.084724
Vary: Origin
Transfer-Encoding: chunked

{"user":{"id":4,"email":"amandabanana@yahoo.com","token":"BAhJIiU5OWE3NWM2YTFlYThiZWNhMDZlNGQxMjVmYWYwZWQzZAY6BkVG--e88e04c3fea071ac5cfe2379c170204d2991c171"}}~/wdi/projects/birthday-api (birthdays)
$  TOKEN=BAhJIiU5OWE3NWM2YTFlYThiZWNhMDZlNGQxMjVmYWYwZWQzZAY6BkVG--e88e04c3fea071ac5cfe2379c170204d2991c171
~/wdi/projects/birthday-api (birthdays)
$ API="${API_ORIGIN:-http://localhost:4741}"
~/wdi/projects/birthday-api (birthdays)
$ URL_PATH="/birthdays"
~/wdi/projects/birthday-api (birthdays)
$ curl "${API}${URL_PATH}" \
   --include \
   --request POST \
   --header "Content-Type: application/json" \
   --header "Authorization: Token token=$TOKEN" \
   --data '{
     "birthday": {
       "born_on":"1990-02-15",
       "family_name":"last name",
       "given_name":"first name",
       "nickname":"nickname"
     }
  }'
HTTP/1.1 422 Unprocessable Entity
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Type: application/json; charset=utf-8
Cache-Control: no-cache
X-Request-Id: 415caced-c01c-41c5-9550-fccae9320954
X-Runtime: 0.006579
Vary: Origin
Transfer-Encoding: chunked

{"users":["must exist"]}~/wdi/projects/birthday-api (birthdays)
$ 
```

I've looked on my db and I can see that my user has been created on the table when I run the curl request and that user is a FK on the birthdays table.

My birthdays controller code looks like this:
```
def create
    # @birthday = Birthday.new(birthday_params)

    @birthday = current_user.birthdays.build(birthday_params)

    if @birthday.save
    render json: @birthday, status: :created
    else
    render json: @birthday.errors, status: :unprocessable_entity
    end

  end
```

My birthdays model: 
```
class Birthday < ApplicationRecord
  belongs_to :user

  validates :born_on, presence: true
  validates :given_name, presence: true
end
```
My user model:
```
class User < ApplicationRecord
  include Authentication
  # has_many :examples
  has_many :birthdays
end
```

I'm not sure what else to check next? I think the code doesn't know how to relate the token I am passing in to the birthday object I am creating, but I'm not sure how to fix it. 

here is some of the error message:
```
{"status":500,"error":"Internal Server Error","exception":"#\u003cNoMethodError: undefined method `users' for #\u003cBirthday:0x007fb81c0d2178\u003e\nDid you mean?  user\n               user=\u003e","traces":{"Application Trace":[{"id":61,"trace":"app/controllers/birthdays_controller.rb:23:in `create'"}],"Framework Trace":[{"id":0,"trace":"activemodel (5.0.1) lib/active_model/attribute_methods.rb:433:in `method_missing'"},{"id":1,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer.rb:194:in `read_attribute_for_serialization'"},{"id":2,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/field.rb:23:in `value'"},{"id":3,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/attributes.rb:22:in `block in attributes'"},{"id":4,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/attributes.rb:19:in `each'"},{"id":5,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/attributes.rb:19:in `each_with_object'"},{"id":6,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/attributes.rb:19:in `attributes'"},{"id":7,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/caching.rb:216:in `fetch_attributes'"},{"id":8,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer.rb:166:in `}
```

**Comment by payne-chris-r** [comment]
Can you post your ERD?
Do users have many birthdays?

**Comment by aeabbott** [comment]
Here's a link to the my ERD: http://imgur.com/a/V9tjL

I can see that birthdays are getting created with the user_id on my birthday table when I send the curl request as well: SELECT * FROM birthdays where user_id = 4; http://imgur.com/LhS7hx4

So I'm thinking now that my code does know how to relate the token, but I'm just not sure why I keep getting the huge error

**Comment by jordanallain** [comment]
can you post your entire birthdays controller?

**Comment by danman01** [comment]
```
{"status":500,"error":"Internal Server Error","exception":"#\u003cNoMethodError: undefined method `users' for #\u003cBirthday:0x007fb81c0d2178\u003e\nDid you mean?  user\n               user=\u003e","traces":{"Application Trace":[{"id":61,"trace":"app/controllers/birthdays_controller.rb:23:in `create'"}],"Framework Trace":[{"id":0,"trace":"activemodel (5.0.1) lib/active_model/attribute_methods.rb:433:in `method_missing'"},{"id":1,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer.rb:194:in `read_attribute_for_serialization'"},{"id":2,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/field.rb:23:in `value'"},{"id":3,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/attributes.rb:22:in `block in attributes'"},{"id":4,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/attributes.rb:19:in `each'"},{"id":5,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/attributes.rb:19:in `each_with_object'"},{"id":6,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/attributes.rb:19:in `attributes'"},{"id":7,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer/concerns/caching.rb:216:in `fetch_attributes'"},{"id":8,"trace":"active_model_serializers (0.10.4) lib/active_model/serializer.rb:166:in `}
```
At some point you're calling `users` on birthday, and this error says, `undefined method users. Did you mean user?` And look down the stack trace, starting at the top, checking each one of those files to trace this error.

Also,
```
{"users":["must exist"]}
```
Is returned from your create Birthdays curl request. Do you have a validation somewhere that is failing? I would try searching your project for calls to `users` and see what you can find, starting with files related to Birthday.

**Comment by aeabbott** [comment]
I figured it out!!! I had the following in my serializer and the :user was causing the issue:
```
class BirthdaySerializer < ActiveModel::Serializer
  attributes :id, :given_name, :family_name, :nickname, :born_on, :user
```
I wanted to see the user id when looked at all my birthdays via the browser so I added users to the serializer, but i think because the user class has extra permissions on it for who can see the users- it was causing my the 500 error when I would try to create.

Using what Danny had written above about looking at the stack trace, I saw that the active_model_serializers was appearing over and over again in the error so I checked out the birthday serializer and found out where users was being called.

Thanks everyone for helping!
